### PR TITLE
Workaround combinations of `-std=c++17` and `-fmodules` caused crash

### PR DESCRIPTION
**Pull request by liuliu**
On Linux (Ubuntu 20.04), SwiftPM will error out with https://github.com/liuliu/swift-mujoco/runs/6837193244?check_suite_focus=true

The crash reported to llvm-project: https://github.com/llvm/llvm-project/issues/55980

This leads me to investigate how and why these flags added. It turns out
it is already disabled on Apple platform with C++.

This PR disabled `-fmodules` for C++ at all because these ongoing
issues. I am not sure if Apple internally benefits with performance
improvements of `-fmodules` because it is dsiabled on their respective
platforms.

Alternatively, I can disable this for c++17 only. Let me know what do
you prefer. Personally I would prefer smaller combinations otherwise you
will have increasingly larger test matrix.

Test Plan:

Checkout https://github.com/liuliu/swift-mujoco on Linux with up-to-date
toolchain, run `swift build`.

Before this PR, it will crash with LLVM reporting instructions.

After this PR, it builds successfully.

### CODE REVIEW

**Comment by neonichu** [comment]
So this is an issue specific to C++17 and works with both older and newer language versions, is that correct? I think in that case it does seem better to limit this to that specific version, though it would also be interesting whether we can enable modules on Darwin these days.

**Comment by liuliu** [comment]
I believe (on my system), both `c++20` and `c++2b` will crash. Seems related to new capabilities added to `c++17` that at odds with `-fmodules`

**Comment by neonichu** [comment]
OK, in this case, I think just disabling it entirely is probably the right thing. We may need to bring the tools-version into it, though, there's probably potential to break existing packages which rely on `-fmodules` being passed on older versions.

**Comment by liuliu** [comment]
Cool. Let me know what I can do on my side (need to do some digging how to bring the tools-version into it).

TBH, the `-fmodules` option caused some other problems for me in the beginning of this week too (following our Slack discussion in the beginning of this week):

 * I structured the C++ libraries dependencies as target dependencies in SwiftPM;
 * The leaf target included one of the dependency does a `#define SOME_FLAG` in the source code, which affects the behavior of that header in the parent target;
 * Because `-fmodules` compiled the parent target separately, and `#include <that header>` will be translate as module import, the compilation result is not what I expected;

Fortunately, this is easy to discover as that `SOME_FLAG` means whether to include the implementation, and by link time, the missing implementation is quickly discovered and can be resolved easily (by not having dependency at all, since it is a C++ fancy header-only lib anyway).

**Comment by neonichu** [comment]
It looks like the clang target already knows about the `toolsVersion` of the containing package, so I think what you could do is retain the previous behavior for tools versions < 5.7 and only apply the new behavior otherwise. This ensures that older packages continue to build the same way they did when they were originally created.

This does presuppose that we would cherry pick this into 5.7 which seems reasonable to me, wdyt @tomerd?

**Comment by liuliu** [comment]
Updated to reflect the latest discussion on `toolsVersion`. Using `toolsVersion` to gate this change to `v5_7` and above.

**Comment by tomerd** [comment]
@compnerd could you also take a look here?

**Comment by tomerd** [comment]
> This does presuppose that we would cherry pick this into 5.7 which seems reasonable to me, wdyt @tomerd?

sounds reasonable

**Comment by compnerd** [comment]
The solution I think is reasonable - disabling modules.  I wonder if there is a better way to write this logic though.  It effectively comes down to enable modules on Linux always when building C, and on Darwin on >= 5.7.  Since 5.7 hasn't been released yet, I really wonder if we should simplify this further to: enable modules for C on Linux only.  But at that point, I start wondering if we want to even consider going further and limiting it to <=5.6.

Thanks for the report on LLVM btw.

**Comment by tomerd** [comment]
a unit test would be good. you can follow some of the example in `MiscellaneousTestCase` which use fixtures to test similar things

**Comment by liuliu** [comment]
Added test case for this crash. Distilled the test case a bit, but not as small as people might like though.

**Comment by liuliu** [comment]
Let me know if there are more things need to be done for this PR. Thanks!

**Comment by neonichu** [comment]
Thanks, @liuliu, nothing is needed from you right now.

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Review by neonichu — APPROVED** [review]

**Comment by neonichu** [comment]
I think at this point it is too late to include this for the 5.7 release (cc @tomerd in case he thinks otherwise), so we'd need to adjust the condition, but otherwise this looks good to merge.

**Comment by liuliu** [comment]
There doesn't seem to be a ToolsVersion.v5_8 in the main branch. Should I add one?

**Comment by neonichu** [comment]
Ah yes, sorry. We can use `vNext` for now and replace it once we know what the next version is going to be.

File: Fixtures/Miscellaneous/CXX17CompilerCrash/v5_8/Package.swift [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// swift-tools-version:999.0.0
```
Comment (liuliu) [inline Fixtures/Miscellaneous/CXX17CompilerCrash/v5_8/Package.swift:1]: Done. The only thing is, when we have v5_8, we need to update this as well.

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Pull request merged**